### PR TITLE
Make GitExtensions project authority of GIT info

### PR DESCRIPTION
**Pull request by vbjay**
Prevents environment data from getting stale by a commit but GitGUI not being rebuilt

Fixes Issue noted in https://github.com/gitextensions/gitextensions/pull/5640#issuecomment-434284350

Changes proposed in this pull request:
- Make GitExtensions project authority of Git info.  Any time  GitExtensions is built git info is recalculated.  
 
Screenshots before and after (if PR changes UI):

What did I do to test the code and ensure quality:
- Ran project and made sure info was still current. 

Has been tested on (remove any that don't apply):

### CODE REVIEW

**Comment by vbjay** [comment]
Sure. I can do that. The issue is still that gitgui doesn't always get
built. See how this package works is adds a file on obj/. The issue is we
want gitextensions to be the authority or the info. It sends in the data
and gitgui displays it. I don't mind adding the Nuget there. It just won't
solve the problem I am solving here.

**Comment by vbjay** [comment]
Nope.  This does not work.  Have to have it in the project file.

**Comment by vbjay** [comment]
@drewnoakes  I can tell you from past experience.  If I added the nuget package to all projects you and everyone else would complain about breadth.  So this solves the issue of stale info not being in the information.

File: GitExtensions/Program.cs [hunk 1]
```diff
@@ -54,6 +54,11 @@ private static void Main()
                 }
             }
 
+            // This is done here so these values can be used in the GitGui project but this project is the authority of the values.
+            UserEnvironmentInformation.Sha = ThisAssembly.Git.Sha;
+            UserEnvironmentInformation.Dirty = ThisAssembly.Git.IsDirty;
+            UserEnvironmentInformation.LockValues();
```
Comment (RussKie) [inline GitExtensions/Program.cs]: I'd change the API to something like:

```
UserEnvironmentInformation.Initialise(ThisAssembly.Git.Sha, ThisAssembly.Git.IsDirty);
```

```
        private static bool _alreadySet;
        private static bool _dirty;
        private static string _sha;

    public static void Initialise(string sha, bool isDirty)
    {
        if (_alreadySet) { return;  // or throw? }
        _alreadySet = true;
        _sha = sha;
        _dirty = isDirty;
    }

    public static string GetInformation()
    {
        if (!_alreadySet) { throw new InvalidOperationException($"{nameof(Init)} must be called first"); }
```

This way the public API surface remains minimal.

Comment (vbjay) [reply GitExtensions/Program.cs]: Done.

File: GitUI/UserEnvironmentInformation.cs [hunk 2]
```diff
@@ -27,12 +33,53 @@ public static string GetInformation()
...
+            {
+                return;
+            }
+
+            _alreadySet = true;
+            _sha = sha;
+            _dirty = isDirty;
+        }
+
+        public static bool Dirty
```
Comment (gerhardol) [inline GitUI/UserEnvironmentInformation.cs]: The Dirty, Sha properties can be removed

Comment (vbjay) [reply GitUI/UserEnvironmentInformation.cs]: Durp. Thanks. I spent too long in front of the computer. Should have seen that.

**Comment by RussKie** [comment]
Unit tests are failing - https://ci.appveyor.com/project/gitextensions/gitextensions/builds/19930317#L225
We need to call `UserEnvironmentInformation.Initialise` from `TranslationTest.cs`

**Comment by vbjay** [comment]
Oh right. Sorry.   I'll fix it.

**Pull request merged**